pkg/scheduler/plugins/coescheduling/core/core.go

```diff
@@ -0,0 +1,490 @@
+					namespacedName := GetNamespaceSplicingName(pod.Namespace, pod.Name)
+					s.Map[namespacedName] = pod
+					klog.Infof("ActivateSiblings add pod's key to PodsToActivate map, pod :%v ", namespacedName)
+				}
+				s.Unlock()
+			}
+		}
+	}
+}
+
+// PreFilter
+// if non-strict-mode, we only do step1 and step2:
+// i.Check whether children in Gang has met the requirements of minimum number under each Gang, and reject the pod if negative.
+// ii.Check whether the Gang is inited, and reject the pod if positive.
+// iii.Check whether the Gang has met the scheduleCycleValid check, and reject the pod if negative.
+// iv.Try update scheduleCycle, scheduleCycleValid, childrenScheduleRoundMap as mentioned above.
+func (pgMgr *PodGroupManager) PreFilter(ctx context.Context, pod *corev1.Pod) error {
+	if !IsPodNeedGang(pod) {
+		return nil
+	}
+	gang := pgMgr.GetGangByPod(pod)
+	if gang == nil {
+		return fmt.Errorf("can't find gang, gangName: %v, podName: %v", GetNamespaceSplicingName(pod.Namespace, GetGangNameByPod(pod)),
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+	}
+	// first try update the global cycle of gang
+	gang.trySetScheduleCycleValid()
+
+	gangScheduleCycle := gang.getScheduleCycle()
+	defer func() {
+		gang.setChildScheduleCycle(pod, gangScheduleCycle)
+	}()
+	// check if gang is inited
+	if !gang.HasGangInit {
+		return fmt.Errorf("gang has not init, gangName: %v, podName: %v", gang.Name,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+	}
+	// check minNum
+	if gang.getChildrenNum() < gang.getGangMinNum() {
+		return fmt.Errorf("gang child pod not collect enough, gangName: %v, podName: %v", gang.Name,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+	}
+
+	gangMode := gang.getGangMode()
+	if gangMode == extension.GangModeStrict {
+		podScheduleCycle := gang.getChildScheduleCycle(pod)
+		if !gang.isScheduleCycleValid() {
+			return fmt.Errorf("gang scheduleCycle not valid, gangName: %v, podName: %v",
+				gang.Name, GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		}
+		if podScheduleCycle >= gangScheduleCycle {
+			return fmt.Errorf("pod's schedule cycle too large, gangName: %v, podName: %v, podCycle: %v, gangCycle: %v",
+				gang.Name, GetNamespaceSplicingName(pod.Namespace, pod.Name), podScheduleCycle, gangScheduleCycle)
+		}
+	}
+	return nil
+}
+
+// PostFilter
+// i. If strict-mode, we will set scheduleCycleValid to false and release all assumed pods.
+// ii. If non-strict mode, we will do nothing.
+func (pgMgr *PodGroupManager) PostFilter(ctx context.Context, pod *corev1.Pod, handle framework.Handle, pluginName string) (*framework.PostFilterResult, *framework.Status) {
+	if !IsPodNeedGang(pod) {
+		return &framework.PostFilterResult{}, framework.NewStatus(framework.Unschedulable, "")
+	}
+	gang := pgMgr.GetGangByPod(pod)
+	if gang == nil {
+		return &framework.PostFilterResult{}, framework.NewStatus(framework.Unschedulable, fmt.Sprintf("can't find gang, gangName: %v, podName: %v",
```

Comment:
Just log the error inside instead of returning it.. and klog.KObj can help make the code clean.